llumnix/llumlet/llumlet.py

```diff
@@ -98,29 +101,38 @@ def from_args(cls,
         llumlet = engine_class.remote(instance_id, backend_type, migration_config, *args, **kwargs)
         return llumlet
 
-    def migrate_out(self, dst_instance_name: str) -> List[str]:
+    def migrate_out(self, dst_instance_name: str, num_requests: int, strict_pre_migration: bool) -> List[str]:
         try:
-            t0 = time.time()
+            self.update_strict_pre_migration(strict_pre_migration)
             migrate_in_ray_actor = ray.get_actor(dst_instance_name, namespace='llumnix')
             dst_instance_id = dst_instance_name[len("instance_"):]
-            logger.info("{}->{} begin migrate out".format(self.instance_id, dst_instance_id))
-            migrate_out_request = self.migration_scheduler.get_migrate_out_request()
             migrated_request_list = []
-            if migrate_out_request is None:
-                return migrated_request_list
-            status = self.migration_coordinator.migrate_out_multistage(migrate_in_ray_actor, migrate_out_request)
-            if status == MigrationStatus.FINISHED_DONE:
-                ray.get(migrate_in_ray_actor.execute_engine_method.remote("commit_dst_request", migrate_out_request))
-                self.backend_engine.free_src_request(migrate_out_request)
-                migrated_request_list.append(migrate_out_request.request_id)
-                migrate_out_request.stage_timestamps.append(time.time())
-                self.backend_engine.remove_migrating_out_request_last_stage(migrate_out_request)
-            else:
-                ray.get(migrate_in_ray_actor.execute_migration_method.remote("free_dst_pre_alloc_cache", migrate_out_request.request_id))
-            t1 = time.time()
-            logger.info("{}->{} migrate done, migrate request {}, status:{}, len:{} blocks, cost:{} ms" \
-                  .format(self.instance_id, dst_instance_id, migrated_request_list, status, \
-                   sum(migrate_out_request.stage_num_blocks_list), (t1 - t0)*1000))
+            continue_migrate = True
+            while continue_migrate:
+                t0 = time.time()
+                migrate_out_request = self.migration_scheduler.get_migrate_out_request()
+                if migrate_out_request is not None:
+                    logger.info("migrate_out {}".format(migrate_out_request.request_id))
+                if migrate_out_request is None or not self.strict_pre_migration:
+                    return migrated_request_list
+                logger.info("{}->{} begin migrate out {}".format(self.instance_id, dst_instance_id, migrate_out_request.request_id))
+                status = self.migration_coordinator.migrate_out_multistage(migrate_in_ray_actor, migrate_out_request)
+                if status == MigrationStatus.FINISHED_DONE:
+                    migrate_out_request.expected_steps = -1
```

Comment:
It seems that not necessary to set back expected_steps, we will drop migrated out request.